runtime-testsuite/test/org/antlr/v4/test/runtime/BaseRuntimeTest.java

```diff
@@ -354,9 +356,200 @@ public static ErrorQueue antlrOnString(String workdir,
 				}
 			}
 		}
+		writeDescriptors(clazz, descriptors);
 		return descriptors.toArray(new RuntimeTestDescriptor[0]);
 	}
 
+	/** Write descriptor files. */
+	private static void writeDescriptors(Class<?> clazz, List<RuntimeTestDescriptor> descriptors) {
+		new File("/tmp/descriptors").mkdir();
+		String groupName = clazz.getSimpleName();
+		groupName = groupName.replace("Descriptors", "");
+		String groupDir = "/tmp/descriptors/" + groupName;
+		new File(groupDir).mkdir();
+
+		for (RuntimeTestDescriptor d : descriptors) {
+			try {
+				Pair<String,String> g = d.getGrammar();
+				String gname = g.a;
+				String grammar = g.b;
+				String filename = d.getTestName()+".txt";
+				String content = "";
```

Comment:
Please use `StringBuilder` for code with a big amount of concatenation. An ordinary string is bad for performance here.

Reply:
@KvanTTT FYI the java compiler replaces string concatenation by StringBuilder calls, so no need to change anything

Reply:
The compiler uses allocations of `StringBuilder` for every line, it's inefficient. Maybe it works fine for dynamic languages such as JavaScript (https://stackoverflow.com/a/16696775/1046374), Python (https://stackoverflow.com/a/12171382/1046374), but not for static-like C#, Java, C++.

To not be unfounded, I've created and compiled a simple test:

```java
public static String method(String x) {
    String content = "";
    content += "1";
    content += x;
    return content;
}
```

Java <= 8 compile the code in the following way:

```
  public static java.lang.String method(java.lang.String);
    Code:
       0: ldc           #2                  // String
       2: astore_1
       3: new           #3                  // class java/lang/StringBuilder
       6: dup
       7: invokespecial #4                  // Method java/lang/StringBuilder."<init>":()V
      10: aload_1
      11: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      14: ldc           #6                  // String 1
      16: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      19: invokevirtual #7                  // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      22: astore_1
      23: new           #3                  // class java/lang/StringBuilder
      26: dup
      27: invokespecial #4                  // Method java/lang/StringBuilder."<init>":()V
      30: aload_1
      31: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      34: aload_0
      35: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      38: invokevirtual #7                  // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      41: astore_1
      42: aload_1
      43: areturn
```

It uses `StringBuilder` but it allocates a new instance for every line. It's inefficient.

Java >= 11 compile the code in a more effective way using `makeConcatWithConstants`:

```
  public static java.lang.String method(java.lang.String);
    Code:
       0: ldc           #2                  // String
       2: astore_1
       3: aload_1
       4: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;)Ljava/lang/String;
       9: astore_1
      10: aload_1
      11: aload_0
      12: invokedynamic #4,  0              // InvokeDynamic #1:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
      17: astore_1
      18: aload_1
      19: areturn
```

But anyway `makeConcatWithConstants` creates an allocation on every line. It's not so efficient for long concatenation especially if loops are used.

Also, `StringBuilder` is more readable in my opinion and it's more canonical for Java.

Reply:
That's a different pattern from:

```
public static String method(String x) {
    String content = ""
         += "1"
         + = x;
    return content;
}
```

which should be used, rather than StringBuilder

Reply:
It differs from the original pattern with `+=` operator.

Consider the following code:

```java
public static String methodPlusAssign(String x, String y) {
    String content = "";
    content += x;
    content += y;
    return content;
}
public static String methodAssign(String x, String y) {
    String content = "" + x + y;
    return content;
}
```

And their JVM bytecode (Java 11):

```
  public static java.lang.String methodPlusAssign(java.lang.String, java.lang.String);
    Code:
       0: ldc           #2                  // String
       2: astore_2
       3: aload_2
       4: aload_0
       5: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
      10: astore_2
      11: aload_2
      12: aload_1
      13: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
      18: astore_2
      19: aload_2
      20: areturn

  public static java.lang.String methodAssign(java.lang.String, java.lang.String);
    Code:
       0: aload_0
       1: aload_1
       2: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;
       7: astore_2
       8: aload_2
       9: areturn
```

There are two invocations in `methodPlusAssign` and only one in `methodPlus`. Also, using `+` is limited and such optimization does not work for loops.

Reply:
Hi guys. Thank you for the excellent analysis, however, I don't think performance is critical here as we load the descriptor files. The difference between 10 ms and 50 ms is not really important for this testing rig.

@ericvergnaud what do you think of the descriptor mechanism so far?

```
[grammar]
grammar T;
a : ID+ {
<writeln("$text")>
};
ID : 'a'..'z'+;
WS : (' '|'\n') -> skip;

[grammarName]
T

[start]
a

[input]
a b c

[output]
"""abc
"""
```

Reply:
I like it, LGTM.
Missing [slaveGrammars] ?
How about adding test name and base class name, such that we can generate test code from the test suite ?

Reply:
BTW I'm off on vacation 15th - 19th december.